core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1279,7 +1283,7 @@ class Partition(val topicPartition: TopicPartition,
   }
 
   def appendRecordsToLeader(records: MemoryRecords, origin: AppendOrigin, requiredAcks: Int,
-                            requestLocal: RequestLocal): LogAppendInfo = {
+                            requestLocal: RequestLocal, verificationGuard: Object = null): LogAppendInfo = {
```

Comment:
Hmm, why are we passing `Object` here and various other places? This is generally an anti-pattern since _anything_ can be passed even though it may not be valid. You should create a type for the verification guard.

Reply:
And having `null` as the default makes it even more unsafe.

Reply:
We don't use object content, we just use object identity (i.e. its address), so passing anything is perfectly fine - 2 different objects would have 2 different addresses.  `null` works because we know that it is not equal to any real object identity.

The scenario that we're solving is this: we need to check some condition under lock, then we need to release the lock and do a bunch of work (that includes sending RPCs and etc.), then once the work is done, we need to make sure that the condition still holds.  So to implement the solution, we create an object and remember a reference to it.  If the condition is changed, the reference is set to `null`.  So when we come back, we check if the reference is still set to the object that we remember, if it's not then we've got raced and the condition isn't valid anymore.  Naively, it may seem that we can use just a Boolean flag, and set it to `true` when we start validation, and if the condition is changed, we set the flag to `false`.  This approach, however, is prone to ABA problem.

There are some details here: https://github.com/apache/kafka/pull/13787#discussion_r1213675427

Reply:
I suggested creating a type, but Artem argued that it was smaller and easier to just use Object.

Reply:
@artemlivshits Note that this is Java, so you actually cannot pass an address (many GC implementations move objects and update their addresses).

> so passing anything is perfectly fine - 2 different objects would have 2 different addresses. null works because we know that it is not equal to any real object identity.

Actually this is not true, if you pass "foo" this is an interned string (just as one example).

Reply:
Ok, maybe it was confusing to refer to object identity as its address, let me make a more precise statement: 2 independently allocated objects cannot reside at the same address at the same point in time, so `new Object != new Object` is true at any given point in time, no matter what GC does, and that's the property that's used in this change.

I don't think there is an expectation that the string literals would be different objects, this code allocates new objects to get unique identity.

Reply:
> I don't think there is an expectation that the string literals would be different objects, this code allocates new objects to get unique identity.

Right, but the lack of strong typing doesn't prevent the wrong instance from being passed. That's why we should not use `Object` like that. In practice, it results in avoidable bugs.